src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs

```diff
@@ -0,0 +1,440 @@
+
+The meaning of a Right value is that the substitution holds IF the predicate
+holds.
+
+TODO: This is different from unification's meaning, so we should either
+convert all bottoms to Left, or we should do it selectively. Doing
+it selectively is not trivial, e.g. a bottom inside a function should become
+Left, but inside a constructor we may be able to keep it as bottom.
+-}
+matchAsUnification
+    ::  ( Hashable variable
+        , FreshVariable variable
+        , MetaOrObject level
+        , Ord (variable level)
+        , Ord (variable Object)
+        , Ord (variable Meta)
+        , Show (variable level)
+        , Show (variable Object)
+        , Show (variable Meta)
+        , SortedVariable variable
+        )
+    => MetadataTools level StepperAttributes
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> Either
+        (UnificationError level)
+        (Counter
+            ( PredicateSubstitution level variable
+            , UnificationProof level variable
+            )
+        )
+matchAsUnification tools first second =
+    case match tools Map.empty first second of
+        Nothing -> Left UnsupportedPatterns
+        Just result -> return $ do  -- Counter monad
+            predicateSubstitution <- result
+            return (predicateSubstitution, EmptyUnificationProof)
+
+match
+    ::  ( Hashable variable
+        , FreshVariable variable
+        , MetaOrObject level
+        , Ord (variable level)
+        , Ord (variable Object)
+        , Ord (variable Meta)
+        , Show (variable level)
+        , Show (variable Object)
+        , Show (variable Meta)
+        , SortedVariable variable
+        )
+    => MetadataTools level StepperAttributes
+    -> Map.Map (variable level) (variable level)
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> Maybe (Counter (PredicateSubstitution level variable))
+match tools quantifiedVariables first second =
+    firstJust
+        [ matchEqualHeadPatterns tools quantifiedVariables first second
+        , matchVariableFunction tools quantifiedVariables first second
+        , matchNonVarToPattern tools first second
+        ]
+  where
+    firstJust :: [Maybe a] -> Maybe a
+    firstJust = listToMaybe . catMaybes
+
+matchEqualHeadPatterns
+    :: ( Show (variable level)
+       , SortedVariable variable
+       , MetaOrObject level
+       , Ord (variable level)
+       , Ord (variable Meta)
+       , Ord (variable Object)
+       , Show (variable Meta)
+       , Show (variable Object)
+       , FreshVariable variable
+       , Hashable variable
+       )
+    => MetadataTools level StepperAttributes
+    -> Map.Map (variable level) (variable level)
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> Maybe (Counter (PredicateSubstitution level variable))
+matchEqualHeadPatterns tools quantifiedVariables first second =
```

Comment:
Hm, for everything that is part of MLPatternClass I could have checked that their `getPatternType` is the same, then I could have zipped and merged their `getPatternChildren`. Would that be better?

With the current way, type checking makes the implementation safer, especially in the unlikely case we're adding more pattern types, but it becomes less obvious that some cases (e.g. application) need special handling because they are hidden among the MLPatternClass patterns.

Reply:
@traiansf @ttuegel BTW, I'll merge the PR as-is, but I'll implement any suggestions on this topic in a follow-up PR.